File: benchmarks/bench_array_hash.pl
#!/usr/bin/perl
use strict;
use warnings;

# Push 2,000,000 integers into an array
my @arr;
my $i = 0;
while ($i < 2000000) {
    push(@arr, $i);
    $i++;
}
print "array size: " . scalar(@arr) . "\n";

# Sum every 100th element to verify
my $sum = 0;
my $j = 0;
while ($j < 2000000) {
    $sum += $arr[$j];
    $j += 100;
}
print "array checksum: $sum\n";

# Insert 500,000 key-value pairs into a hash
my %h;
my $k = 0;
while ($k < 500000) {
    $h{"key" . $k} = $k;
    $k++;
}
print "hash size: " . scalar(keys(%h)) . "\n";

# Look up all 500,000 values
my $lookup_sum = 0;
my $m = 0;
while ($m < 500000) {
    $lookup_sum += $h{"key" . $m};
    $m++;
}
print "lookup sum: $lookup_sum\n";

# Delete all keys
my $n = 0;
while ($n < 500000) {
    delete($h{"key" . $n});
    $n++;
}
print "after delete: " . scalar(keys(%h)) . "\n";

File: benchmarks/bench_array_hash.rb
#!/usr/bin/env ruby

# Push 2,000,000 integers into an array
arr = []
2_000_000.times { |i| arr.push(i) }
puts "array size: #{arr.length}"

# Sum every 100th element to verify
s = 0
(0...2_000_000).step(100) { |j| s += arr[j] }
puts "array checksum: #{s}"

# Insert 500,000 key-value pairs into a hash
h = {}
500_000.times { |k| h["key#{k}"] = k }
puts "hash size: #{h.length}"

# Look up all 500,000 values
lookup_sum = 0
500_000.times { |m| lookup_sum += h["key#{m}"] }
puts "lookup sum: #{lookup_sum}"

# Delete all keys
500_000.times { |n| h.delete("key#{n}") }
puts "after delete: #{h.length}"
